File: 301-400/0395-Longest-Substring/0395.py
class Solution(object):
  def longestSubstring(self, s, k):
    if len(s) < k:
      return 0
    for c in set(s):
      if s.count(c) < k:
        return max(self.longestSubstring(t, k) for t in s.split(c))
    return len(s)

File: 301-400/0395-Longest-Substring/0395.go
func longestSubstring(s string, k int) (ans int) {
	if s == "" {
		return
	}

	cnt := [26]int{}
	for _, ch := range s {
		cnt[ch-'a']++
	}

	var split byte
	for i, c := range cnt[:] {
		if 0 < c && c < k {
			split = 'a' + byte(i)
			break
		}
	}
	if split == 0 {
		return len(s)
	}
	for _, subStr := range strings.Split(s, string(split)) {
		ans = max(ans, longestSubstring(subStr, k))
	}
	return
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}